[Anchor] Add PVP damage multiplier

flee135 opened the pull request:
Adds dropdown in anchor menu for admin to adjust PvP damage multipliers. Defaults to 2x, which is the current hardcoded behavior.

Could use feedback on the text descriptions for each multiplier. Not sure if it's too lengthy, unnecessary to include, or if it belongs in a tooltip. 

<img width="1378" height="658" alt="Screenshot 2025-12-08 192906" src="https://github.com/user-attachments/assets/7c247963-073a-4b97-8033-da415a3ca787" />


<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4862932349.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4862951210.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4863129006.zip)
<!--- section:artifacts:end -->

flee135 committed:
Add PVP damage multiplier

garrettjoecox commented on soh/soh/Network/Anchor/Menu.cpp:
```diff
@@ -10,6 +10,17 @@ extern std::shared_ptr<AnchorRoomWindow> mAnchorRoomWindow;
+    "16x: Kokiri sword slash damages four hearts.",
+    "32x: Kokiri sword slash damages eight hearts.",
+    "64x: Kokiri sword slash damages 16 hearts.",
+    "128x: Kokiri sword slash damages 16 hearts through double defense.",
```
I think we could reduce the amount of options here, probably 1, 2, 4, and 1 hit ko

flee135 replied:
Agreed that some of these options are unnecessary. I think 8x is still meaningful (and is how I used to play). It helps make ranged fights not drag out for too long.

WDYT about 1, 2, 4, 8, and OHKO?

flee135 replied:
Dropdown now looks like this.

<img width="688" height="428" alt="image" src="https://github.com/user-attachments/assets/4263cc0f-c8fa-4b0f-8d85-dac22ffb3ded" />

flee135 committed:
Remove high multipliers

garrettjoecox commented on soh/soh/Network/Anchor/Packets/UpdateRoomState.cpp:
```diff
@@ -49,6 +51,7 @@ void Anchor::HandlePacket_UpdateRoomState(nlohmann::json payload) {
 
     roomState.ownerClientId = payload["state"]["ownerClientId"].get<uint32_t>();
     roomState.pvpMode = payload["state"]["pvpMode"].get<u8>();
+    roomState.pvpDamageMult = payload["state"]["pvpDamageMult"].get<u8>();
```
If this client joins a room that is hosted by an older client (without this setting) this will result in a crash. Could you do a conditional check here? I think there is probably some examples somewhere, if not here then in the older anchor code

flee135 replied:
Thanks, good catch. Changed it to not explicitly using a conditional, but instead calling .value() to use value of pvpDamageMult if it exists in the payload, or a default of 1 if it doesn't, which should correspond to the old 2x behavior.

garrettjoecox replied:
Good find, I'll probably migrate to using this in a few other places as well.

garrettjoecox reviewed (approved):
Pending the backwards compatibility check LGTM

flee135 committed:
Conditional pvpDamageMult

Call .value() to use pvpDamageMult if present in UpdateRoomState packet, or default to 1. Prevents crashes when connecting with old clients without the new field.

Malkierian commented:
I'd rather not have such long texts inside the dropdown like that. Not sure how it could be done, since a tooltip would get pretty big, too, but such long lines would cause issues for smaller screens anyway. Tooltip, at least, would be better than it is now, though.

garrettjoecox commented:
> I'd rather not have such long texts inside the dropdown like that. Not sure how it could be done, since a tooltip would get pretty big, too, but such long lines would cause issues for smaller screens anyway. Tooltip, at least, would be better than it is now, though.

They are shorter than the screenshot after the options were reduced, what do you think?

Malkierian commented:
Ah, yeah, that's much smaller than the screenshot, which was what I was working off of.

flee135 commented:
> I'd rather not have such long texts inside the dropdown like that. Not sure how it could be done, since a tooltip would get pretty big, too, but such long lines would cause issues for smaller screens anyway. Tooltip, at least, would be better than it is now, though.

Whoops, sorry I missed this, I'm currently traveling for the holidays and don't have a machine setup to build ship right now. If it's still a problem, I can try to set that up and try moving it to a tooltip and have the dropdown only include the multipliers. Or I can revisit this in mid-January when I'm back from travel.

serprex commented on soh/soh/Network/Anchor/Packets/DamagePlayer.cpp:
```diff
@@ -47,7 +66,7 @@ void Anchor::HandlePacket_DamagePlayer(nlohmann::json payload) {
     u8 damage = payload["damage"].get<u8>();
 
-    self->actor.colChkInfo.damage = damage * 8; // Arbitrary number currently, need to fine tune
+    self->actor.colChkInfo.damage = damage * GetPvpDamageMultiplier(roomState) * 4;
```
this is a u8, with OHKO you're going to overflow it

passing damage as parameter to `GetPvpDamageMultiplier` you could have OHKO ignore damage value & return 63 (need to avoid overflow from `* 4`). But I'm also thinking might as well push the `* 4` down too, renaming `GetPvpDamageMultiplier` to `TransformPvpDamage`
